[Form] [TwigBridge] Added option to disable usage of default themes when rendering a form

emodric opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/8495

This adds a possibility to use `only` keyword in `form_theme` tag to disable usage of globally defined form themes, e.g.

`{% form_theme form with ['common.html.twig', 'form/fields.html.twig'] only %}`

Otherwise, in order to completely control the rendering of the forms (for example in custom admin interfaces), one would need to use a form theme which has all the possible twig blocks defined to prevent globally defined themes to interfere with the rendering.

`only` keyword is already used when including a Twig template to transfer only the variables which are explicitly defined in the `include` tag, so it seemed natural to use it here too.

This, of course, means that the user will need to manually `use` all of the templates that are required to render the form, including `form_div_layout.html.twig`

This issue is described in details over at Symfony Demo repo: https://github.com/symfony/symfony-demo/issues/515

TODO:

- [x] submit changes to the documentation

emodric committed:
[Form] [TwigBridge] Added option to disable usage of default themes when rendering a form

HeahDude commented on src/Symfony/Bridge/Twig/Node/FormThemeNode.php:
```diff
@@ -16,9 +16,9 @@
 class FormThemeNode extends \Twig_Node
 {
-    public function __construct(\Twig_Node $form, \Twig_Node $resources, $lineno, $tag = null)
+    public function __construct(\Twig_Node $form, \Twig_Node $resources, $only, $lineno, $tag = null)
```
Should be added as last argument with default `false` to keep BC.

sstok replied:
Isn't class considered an internal detail?

HeahDude replied:
Actually it's not marked as is, and anyone could extend the parser and the node to use custom implementation.

I don't know if it's a a doable/valid use case, but this API is public for now anyway, so BC should be preserved.

emodric replied:
@HeahDude Will do! As I am currently on vacation, give me couple of weeks and I'll do it.

emodric replied:
@HeahDude Placed the `$only` argument at the end and rebased on top of `3.4`.

emodric commented:
Hi, what is the status PR? Do you think it could end up in 3.4?

stof commented on src/Symfony/Component/Form/FormRendererEngineInterface.php:
```diff
@@ -21,11 +21,13 @@
+     *                                   default themes to render the view
      */
-    public function setTheme(FormView $view, $themes);
+    public function setTheme(FormView $view, $themes, $useDefaultThemes = true);
```
adding the argument in the interface is a BC break. So this cannot be done before 4.0 (and 3.4 should trigger a deprecation warning in case it finds an implementation of the interface without the extra optional attribute)

emodric replied:
Okay, makes sense. How should I proceed then?

stof commented on src/Symfony/Component/Form/FormRendererInterface.php:
```diff
@@ -28,11 +28,13 @@ public function getEngine();
+     *                                   default themes to render the view
      */
-    public function setTheme(FormView $view, $themes);
+    public function setTheme(FormView $view, $themes, $useDefaultThemes = true);
```
same here

fabpot commented:
@emodric I like this feature, can you implement the suggestions made by @stof?

emodric commented:
@fabpot Sure, but how do you want me to handle it?

Open a new PR for deprecations only against 3.4 and target this one for master or did you have something else in mind?

fabpot commented:
You need to remove the argument and use `get_func_args()` to determine how it was called to handle the new arg and trigger a dep notice if needed. You can find quite a few examples in the code base doing exactly this.

emodric commented:
Ah okay, I figured you wanted the feature to go in only in 4.0. I'll do it first thing tomorrow morning.

emodric commented:
@fabpot I've updated the code to include requested deprecations. I updated the tests too in order not to trigger deprecation notices in them. There is a test failure in Twig bridge tests, but I don't see how it is related to this change.

nicolas-grekas reviewed (approved):
(failure unrelated)

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Templating/Helper/FormHelper.php:
```diff
@@ -54,7 +54,12 @@ public function getName()
-        $this->renderer->setTheme($view, $themes);
+        $args = func_get_args();
+        if (!isset($args[2])) {
+            @trigger_error(sprintf('Calling setTheme method of %s without the third boolean argument $useDefaultThemes is deprecated since 3.4 and will be removed in 4.0. Call the method with the third argument included and use boolean true to keep current behaviour.', self::class), E_USER_DEPRECATED);
```
I don't understand why we would want to force people to pass a third argument, which should be `true` by default.

emodric replied:
Hm.. I'm not sure what exactly deprecation message should be then if the third argument is optional? Can't we just use it if it's there, and if not, default to `true`?

fabpot replied:
I think this is the right approach.

emodric replied:
Okay, will fix it ASAP. Where do you want me to document the fact that in 4.0 the interface is going to receive a third argument? I've seen somewhere in the codebase the usage of an inline comment in the style of `public function setTheme(FormView $view, $themes /* $useDefaultThemes = true */ );`

fabpot replied:
Yes, that's what we are doing.

emodric replied:
@fabpot Alright, I've updated the code now as per instructions.

nicolas-grekas commented on src/Symfony/Component/Form/FormRendererEngineInterface.php:
```diff
@@ -21,11 +21,13 @@
+     *                                   in the engine, will be added to the interface in 4.0
      */
-    public function setTheme(FormView $view, $themes);
+    public function setTheme(FormView $view, $themes /*, $useDefaultThemes = true */);
```
in all places where this is added, the code should throw a deprecation when the argument is not provided (using func_num_args, see other places in the code base).

emodric replied:
@nicolas-grekas As per @fabpot, deprecation notice is not needed if the argument is optional?

fabpot replied:
I think the current code is good as is :)

fabpot commented:
Thank you @emodric.

emodric commented:
Thanks everyone for review :)
